packages/block-editor/src/hooks/block-binding-support/with-block-binding-support.js

```diff
@@ -0,0 +1,178 @@
+/**
+ * WordPress dependencies
+ */
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { useEffect, useCallback, useRef } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../../store';
+import { useSelect } from '@wordpress/data';
+import { unlock } from '../../../../editor/src/lock-unlock';
+import { getBlockType } from '@wordpress/blocks';
+
+/**
+ * This component is responsible detecting and
+ * propagating data changes between block attribute and
+ * the block-binding source property.
+ *
+ * The app creates an instance of this component for each
+ * pair of block-attribute/source-property.
+ *
+ * @param {Object}   props            - The component props.
+ * @param {string}   props.attrName   - The attribute name.
+ * @param {any}      props.attrValue  - The attribute value.
+ * @param {Function} props.useSource  - The custom hook to use the source.
+ * @param {Object}   props.blockProps - The block props with bound attribute.
+ * @param {Object}   props.args       - The arguments to pass to the source.
+ * @return {null}                       This is a data-handling component. Render nothing.
+ */
+const BlockBindingConnector = ( {
+	args,
+	attrName,
+	attrValue,
+	blockProps,
+	useSource,
+} ) => {
+	const {
+		placeholder,
+		value: propValue,
+		updateValue: updatePropValue,
+	} = useSource( blockProps, args );
+
+	const blockName = blockProps.name;
+
+	const setAttributes = blockProps.setAttributes;
+
+	const updateBoundAttibute = useCallback(
+		( newAttrValue ) => {
+			setAttributes( {
+				[ attrName ]: newAttrValue,
+			} );
+		},
+		[ attrName, setAttributes ]
+	);
+
+	// Store a reference to the last value and attribute value.
+	const lastPropValue = useRef( propValue );
+	const lastAttrValue = useRef( attrValue );
+
+	/*
+	 * Initially sync (first render / onMount ) attribute
+	 * value with the source prop value.
+	 */
+	useEffect( () => {
+		updateBoundAttibute( propValue );
+	}, [ propValue, updateBoundAttibute ] );
+
+	/*
+	 * Sync data.
+	 * This effect will run every time
+	 * the attribute value or the prop value changes.
+	 * It will sync them in both directions.
+	 */
+	useEffect( () => {
+		/*
+		 * Source Prop => Block Attribute
+		 *
+		 * Detect changes in source prop value,
+		 * and update the attribute value accordingly.
+		 */
+		if ( typeof propValue !== 'undefined' ) {
+			if ( propValue !== lastPropValue.current ) {
```

Comment:
We should check here if the `propValue` or the `lastPropValue.current` is an instance of `RichTextData` because it might be the case for Paragraph blocks.

If so, we'd have to get the value from RichTextData instance with `.text()` and compare those values.

Reply:
Is it something that the API should take care of? Should it be agnostic about the type?

The API provides the value to the source handler, no matter the type, and the source handler takes over, in this case, to get raw from the RichTextData instance [here](https://github.com/WordPress/gutenberg/blob/868a0aee3e2ee367b8dabb2bdc8dcb2a3195a70e/packages/editor/src/bindings/post-meta.js#L34-L36).

How does it scale when let's suppose, it connects custom blocks with custom handlers?

I think it's possible to pick the value depending on the attribute type. [Here](https://github.com/WordPress/gutenberg/blob/a1719b779cdb7e4accbd45c96406fb5caf8e5560/packages/block-library/src/paragraph/block.json#L16) we can recognize the attribute type of the core/paragraph block.

But still, I'm asking before proceeding since maybe it does make sense.

Reply:
I've been thinking about it a bit. It really depends on what kind of contract we want to define for the handlers.

I'm leaning towards my original suggestion to add the check here and the [updater function ](https://github.com/WordPress/gutenberg/blob/868a0aee3e2ee367b8dabb2bdc8dcb2a3195a70e/packages/editor/src/bindings/post-meta.js#L34-L36) would then only receive the raw value without any `RichTextData` wrappers.

Ideally, developers who create custom sources should not need to know about `RichTextData` and should just get a value and then do whatever they want with it.

Another idea would be to wrap the value passed to the `updateValue()` handler in an `Event` object of some sort which has an explicit type. Sth like: `Event{ type: "string"|"rich-text"|"number", value: <the string or the RichTextData} `. Somewhat akin to the value that gets passed to `onChange()` in React.

---

All that said, we can leave that discussion for later as we should merge the refactor first, then the work on the UI improvements (https://github.com/WordPress/gutenberg/issues/58978) and then we can come back to editing and attack it more holistically 🙂 .

Reply:
👍 to deal with this in a follow-up.